intelmq/bots/parsers/abusech/parser_ip.py

```diff
@@ -6,36 +6,97 @@
+            'time.source',
+            'source.ip',
+            'source.port',
+            'malware.name'
+        ],
+        'malware': 'Cridex'
+    },
+    'https://zeustracker.abuse.ch/blocklist.php?download=ipblocklist': {
+        'format': [
+            'source.ip'
+        ],
+        'malware': 'Zeus'
+    },
+    'https://zeustracker.abuse.ch/blocklist.php?download=badips': {
+        'format': [
+            'source.ip'
+        ],
+        'malware': 'Zeus'
+    }
+}
 
 
 class AbusechIPParserBot(ParserBot):
     lastgenerated = None
 
+    def parse(self, report: dict):
+        lines = utils.base64_decode(report.get("raw")).splitlines()
+        regex = re.compile(r'^#+.*')  # filter out comments
+        top_line = list(filter(lambda i: not regex.search(i), lines))[0]
+        feed = report['feed.url']
+
+        fields = top_line.split(',')
+        if len(fields) is not len(FEEDS[feed]['format']):
+            self.logger.warning("Feed '{}' has not the expected fields: {} != {}".format(feed,
+                                                                                         len(fields),
+                                                                                         len(FEEDS[feed]['format'])))
+            raise PipelineError("Abusech ip parser is not up to date with the format online")
+
+        for line in utils.base64_decode(report.get("raw")).splitlines():
+            line = line.strip()
+            if not any([line.startswith(prefix) for prefix in self.ignore_lines_starting]):
+                yield line
+
     def parse_line(self, line, report):
         if line.startswith('#'):
             self.tempdata.append(line)
-            if 'Generated on' in line:
-                row = line.strip('# ')[13:]
-                self.lastgenerated = dateutil.parser.parse(row).isoformat()
+            if 'Last updated' in line:
+                self.lastgenerated = dateutil.parser.parse(
+                    re.search('[0-9]{4}.[0-9]{2}.[0-9]{2}.[0-9]{2}.[0-9]{2}.[0-9]{2}( UTC)?', line).group(
+                        0)).isoformat()
         else:
             event = self.new_event(report)
-            event.add('time.source', self.lastgenerated)
-            event.add('classification.taxonomy', 'malicious code')
-            event.add('classification.type', 'c&c')
-            event.add('source.ip', line)
-            event.add("raw", line)
-            event.add("malware.name", SOURCE_FEEDS[report["feed.url"]])
+            self.__process_defaults(event, line, report['feed.url'])
+            self.__process_fields(event, line, report['feed.url'])
             yield event
 
+    def __process_defaults(self, event, line, feed_url):
+        defaults = set([
+            ('malware.name', FEEDS[feed_url]['malware']),
+            ('raw', line),
+            ('classification.type', 'c&c'),
+            ('classification.taxonomy', 'malicious code'),
+            ('time.observation', self.lastgenerated)
```

Comment:
Please remove this one. The observation time is the time when the data has been fetched

Reply:
Should the `time.observation` not reflect the changes within IntelMQ along with all supplied knowledge? When the time of last feed refresh is known, that is the last observable time in the parser's knowledge.

The same logic was applied in [this](https://github.com/certtools/intelmq/blob/fb5eba287ddd4234a3a24eba72d025dfd03079af/intelmq/bots/parsers/cert_eu/parser_csv.py#L46) bot.

If the `time.observation` field modeled the physical observation, repeated reads would produce unique messages, throwing away a piece of important information. For the sake of generic behavior, when you imagine a feed consisting of e.g. only IP addresses and the first line comment that says: `Last updated 2001-02-07`, would it be acceptable to have `time.observation` to state the (current) event read time, discarding the fact that it is an 18 years old information?

Reply:
> When the time of last feed refresh is known, that is the last observable time in the parser's knowledge.

Yeah, but that's not what `time.observation` is for. I now discussed with @aaronkaplan what the intention and use case of the field is and this is the result (which I will also add to the documentation and fix in the cert-eu parser):

We save two timestamps, one is the time of the event itself. The time, when the incident happened, or the port was open, the connection has been made, etc. This is reported by the source, thus `time.source`.
The other useful information is the time, when the event was first processed by the local intelmq instance, the "observation" of the event. As an example, we at cert.at use this time often to search to events which have been processed e.g. today and to enhance the query speed on the database (when searching for certain events where you now that they are from today).
Speaking generally again: What is the advantage of knowing at what time the feed has been last updated by the feed provider? In case of abuse.ch this time is the most recent "firstseen" (time.source) value. I actually don't see a reason why I do need to attach this information to every event of the feed. The `time.source` field already has the time of the actual event (which is the more important one here).

For plain list of IPs with a single "last updated" information: That's the case for blocklists and we do have documentation for a few of them in intelmq too. In that case we know "The IP [source.ip] was on this list as of [time.source/lastgenerated]". An example is the (current) abusech ip parser:
https://github.com/certtools/intelmq/blob/0ba0a8e/intelmq/bots/parsers/abusech/parser_ip.py

Reply:
yes, and if it's really necessary to add a "last_updated" for a specific feed we could add that to extra. Only when that field is so common (like it would be needed for all feeds), then we should have it standardized. 

But yes, standardizing the internal format (DHO) is non trivial.

Reply:
Resulting commits:
33a18883264ecc739716c494dd7917e9f4533234
7f484c3dc